mkosi/kmod.py

```diff
@@ -181,28 +173,21 @@ def gen_required_kernel_modules(
         mods = set((modulesd / "kernel").rglob("*.ko*"))
         firmware = set()
 
-    yield modulesd.parent
-    yield modulesd
-    yield modulesd / "kernel"
-
-    if (root / "usr/lib/firmware").exists():
-        yield root / "usr/lib/firmware"
-
-    yield from {p for m in sorted(mods) for p in parents_relative_to(m, modulesd / "kernel")}
-    yield from {p for f in sorted(firmware) for p in parents_relative_to(f, root / "usr/lib/firmware")}
-
-    for p in itertools.chain(sorted(mods), sorted(firmware)):
-        yield p
+    if not mods and not firmware:
+        return
 
-    for p in (root / modulesd).iterdir():
-        if p.name.startswith("modules"):
-            yield p
+    yield from sorted(
+        itertools.chain(
+            {p for f in mods | firmware for p in list(f.parents)[:list(f.parents).index(root / "usr/lib")]},
+            mods,
+            firmware,
+            modulesd.glob("modules*"),
+        )
+    )
 
-    if (root / modulesd / "vdso").exists():
+    if (modulesd / "vdso").exists():
         yield modulesd / "vdso"
-
-        for p in (root / modulesd / "vdso").iterdir():
-            yield p
+        yield from sorted((modulesd / "vdso").iterdir())
```

Comment:
I am not fully sure what these files to but I think they should be above the `if not mods or not de: return`?

Reply:
Also the sorting should be redundant as we do not return any subdirectories

Reply:
Direction iteration order is not stable, so this is required for reproducible builds as well.